### PR TITLE
Task_3.3

### PR DESCRIPTION
**Pull request by tranthihiep**

### CODE REVIEW

File: app/src/main/java/com/framgia/music/FragmentAlbum.java [hunk 1]
```diff
@@ -79,7 +79,7 @@ public void onClick(View view) {
             @Override
             public void onClick(View view) {
                 if (edtAlbum.getText().toString().equals("")) {
-                    Toast.makeText(getContext(), "Please input album's name ", Toast.LENGTH_SHORT)
+                    Toast.makeText(getContext(),R.string.please_input_album_name, Toast.LENGTH_SHORT)
```
Comment (le-quang-dao) [inline app/src/main/java/com/framgia/music/FragmentAlbum.java]: ERROR: Line is longer than 100 characters (found 101).

Comment (le-quang-dao) [inline app/src/main/java/com/framgia/music/FragmentAlbum.java]: ERROR: ',' is not followed by whitespace.

File: app/src/main/java/com/framgia/music/MainActivity.java [hunk 2]
```diff
@@ -42,5 +51,44 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            if (checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)
+                    != PackageManager.PERMISSION_GRANTED) {
+
+                //Permisson don't granted
+                if (shouldShowRequestPermissionRationale(
+                        Manifest.permission.WRITE_EXTERNAL_STORAGE)) {
+                    Toast.makeText(MainActivity.this, R.string.permission_denied,
+                            Toast.LENGTH_SHORT).show();
+                }
```
Comment (le-quang-dao) [inline app/src/main/java/com/framgia/music/MainActivity.java]: ERROR: '}' at column 17 should be on the same line as the next part of a multi-block statement (one that directly contains multiple blocks: if/else-if/else, do/while or try/catch/finally).

File: app/src/main/java/com/framgia/music/SongDataHelper.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+                    }
+                } else if (file.getName().endsWith(".mp3")) {
+                    HashMap<String, String> song = new HashMap<>();
+                    song.put("file_path", file.getAbsolutePath());
+                    song.put("file_name", file.getName());
+                    fileList.add(song);
+                }
+            }
+            return fileList;
+        } catch (Exception e) {
```
Comment (le-quang-dao) [inline app/src/main/java/com/framgia/music/SongDataHelper.java]: INFO: Catching 'Exception' is not allowed.

File: app/src/main/java/com/framgia/music/SongDataHelper.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    @Override
+    public void onUpgrade(SQLiteDatabase sqLiteDatabase, int i, int i1) {
+
+    }
+
+    public void initListSong() {
+        insertListSongFromStorage();
+        songListFinal = new ArrayList<>();
```
Comment (le-quang-dao) [inline app/src/main/java/com/framgia/music/SongDataHelper.java]: ERROR: [STYLE][ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD] Write to static field com.framgia.music.SongDataHelper.songListFinal from instance method com.framgia.music.SongDataHelper.initListSong()
http://findbugs.sourceforge.net/bugDescriptions.html#ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD

File: app/src/main/java/com/framgia/music/SongDataHelper.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+        songListFinal = getListSongFromDatabase();
```
Comment (le-quang-dao) [inline app/src/main/java/com/framgia/music/SongDataHelper.java]: ERROR: [STYLE][ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD] Write to static field com.framgia.music.SongDataHelper.songListFinal from instance method com.framgia.music.SongDataHelper.initListSong()
http://findbugs.sourceforge.net/bugDescriptions.html#ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD

File: app/src/main/java/com/framgia/music/MainActivity.java [hunk 6]
```diff
@@ -42,5 +51,44 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
+                        Toast.LENGTH_SHORT).show();
+            }
+        } else {
+            super.onRequestPermissionsResult(requestCode, permissions, grantResults);
+        }
+    }
+
+    public void initPermission() {
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
+            if (checkSelfPermission(Manifest.permission.WRITE_EXTERNAL_STORAGE)
```
Comment (le-quang-dao) [inline app/src/main/java/com/framgia/music/MainActivity.java]: WARNING: [CollapsibleIfStatements] These nested if statements could be combined
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/basic.html#CollapsibleIfStatements

File: app/src/main/res/layout/row_song.xml [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+            app:srcCompat="@drawable/favorite" />
+
+        <ImageView
+            android:contentDescription="@string/description"
+            android:id="@+id/btnUnFavRowSong"
+            android:layout_width="@dimen/dp_50"
+            android:layout_height="@dimen/dp_50"
+            app:srcCompat="@drawable/unfavorite" />
+        />
+    </FrameLayout>
```
Comment (le-quang-dao) [inline app/src/main/res/layout/row_song.xml]: WARNING: [ExtraText] Unexpected text found in layout file: "/>"
 Layout resource files should only contain elements and attributes. Any XML text content found \
in the file is likely accidental (and potentially dangerous if the text resembles XML and the \
developer believes the text to be functional)

File: app/src/main/java/com/framgia/music/SongDataHelper.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+import java.io.File;
+import java.util.ArrayList;
+import java.util.HashMap;
+
+/**
+ * Created by trant on 22/01/2018.
+ */
+
+public class SongDataHelper extends SQLiteOpenHelper {
+    private static ArrayList<SongMusic> SONG_LIST_FINAL = null;
```
Comment (le-quang-dao) [inline app/src/main/java/com/framgia/music/SongDataHelper.java]: WARNING: Name 'SONG_LIST_FINAL' must match pattern '^[a-z][a-zA-Z0-9]*$'.

File: app/src/main/java/com/framgia/music/SongDataHelper.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+                        break;
+                    }
+                } else if (file.getName().endsWith(".mp3")) {
+                    HashMap<String, String> song = new HashMap<>();
+                    song.put("file_path", file.getAbsolutePath());
+                    song.put("file_name", file.getName());
+                    fileList.add(song);
+                }
+            }
+        } catch (Exception e) {
```
Comment (le-quang-dao) [inline app/src/main/java/com/framgia/music/SongDataHelper.java]: INFO: Catching 'Exception' is not allowed.

File: app/src/main/java/com/framgia/music/SongDataHelper.java [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+    }
+
+    @Override
+    public void onUpgrade(SQLiteDatabase sqLiteDatabase, int i, int i1) {
+
+    }
+
+    public void initListSong() {
+        insertListSongFromStorage();
+        SONG_LIST_FINAL = getListSongFromDatabase();
```
Comment (le-quang-dao) [inline app/src/main/java/com/framgia/music/SongDataHelper.java]: ERROR: [STYLE][ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD] Write to static field com.framgia.music.SongDataHelper.SONG_LIST_FINAL from instance method com.framgia.music.SongDataHelper.initListSong()
http://findbugs.sourceforge.net/bugDescriptions.html#ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD

File: app/src/main/java/com/framgia/music/SongDataHelper.java (hunk 8 above)
Comment (le-quang-dao) [inline app/src/main/java/com/framgia/music/SongDataHelper.java]: WARNING: [SuspiciousConstantFieldName] The field name indicates a constant but its modifiers do not
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/naming.html#SuspiciousConstantFieldName

File: app/src/main/res/layout/row_song.xml [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+
+        <ImageView
+            android:id="@+id/btnUnFavRowSong"
+            android:layout_width="@dimen/dp_50"
+            android:layout_height="@dimen/dp_50"
+            android:contentDescription="@string/description"
+            app:srcCompat="@drawable/unfavorite"
+            />
+        />
+    </FrameLayout>
```
Comment (le-quang-dao) [inline app/src/main/res/layout/row_song.xml]: WARNING: [ExtraText] Unexpected text found in layout file: "/>"
 Layout resource files should only contain elements and attributes. Any XML text content found \
in the file is likely accidental (and potentially dangerous if the text resembles XML and the \
developer believes the text to be functional)

File: app/src/main/java/com/framgia/music/MainActivity.java [hunk 12]
```diff
@@ -1,18 +1,26 @@
...
 import android.os.Bundle;
+import android.support.annotation.NonNull;
+import android.support.v4.app.ActivityCompat;
 import android.support.v7.app.AppCompatActivity;
 import android.view.Menu;
 import android.view.MenuItem;
 
 public class MainActivity extends AppCompatActivity {
+    private SongDataHelper mSongDataHelper;
+    private final int REQUEST_PERMISSION = 4;
```
Comment (le-quang-dao) [inline app/src/main/java/com/framgia/music/MainActivity.java]: WARNING: Name 'REQUEST_PERMISSION' must match pattern '^[a-z][a-zA-Z0-9]*$'.

File: app/src/main/java/com/framgia/music/MainActivity.java [hunk 13]
```diff
@@ -42,5 +50,26 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
+        if (requestCode == REQUEST_PERMISSION) {
+            initPermission();
+        }
+    }
+
+    private void initPermission() {
+        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE)
+                != PackageManager.PERMISSION_GRANTED) {
+            ActivityCompat.requestPermissions(this,
+                    new String[] { Manifest.permission.WRITE_EXTERNAL_STORAGE }, REQUEST_PERMISSION);
```
Comment (le-quang-dao) [inline app/src/main/java/com/framgia/music/MainActivity.java]: ERROR: Line is longer than 100 characters (found 101).

File: app/src/main/java/com/framgia/music/SongDataHelper.java [hunk 14]
```diff
@@ -0,0 +1,132 @@
...
+                    }
+                } else if (file.getName().endsWith(".mp3")) {
+                    HashMap<String, String> song = new HashMap<>();
+                    song.put("file_path", file.getAbsolutePath());
+                    song.put("file_name", file.getName());
+                    fileList.add(song);
+                }
+            }
+            
+        } catch (Exception e) {
```
Comment (le-quang-dao) [inline app/src/main/java/com/framgia/music/SongDataHelper.java]: INFO: Catching 'Exception' is not allowed.

File: app/src/main/java/com/framgia/music/SongDataHelper.java [hunk 15]
```diff
@@ -0,0 +1,132 @@
...
+            Log.e("Error", "Get list not success ");
+        }
```
Comment (le-quang-dao) [inline app/src/main/java/com/framgia/music/SongDataHelper.java]: ERROR: '}' at column 9 should be on the same line as the next part of a multi-block statement (one that directly contains multiple blocks: if/else-if/else, do/while or try/catch/finally).

File: app/src/main/java/com/framgia/music/SongDataHelper.java [hunk 16]
```diff
@@ -0,0 +1,132 @@
...
+    }
+
+    @Override
+    public void onUpgrade(SQLiteDatabase sqLiteDatabase, int i, int i1) {
+
+    }
+
+    public void initListSong() {
+        insertListSongFromStorage();
+        SongDataHelper.sSongListFinaL = getListSongFromDatabase();
```
Comment (le-quang-dao) [inline app/src/main/java/com/framgia/music/SongDataHelper.java]: ERROR: [STYLE][ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD] Write to static field com.framgia.music.SongDataHelper.sSongListFinaL from instance method com.framgia.music.SongDataHelper.initListSong()
http://findbugs.sourceforge.net/bugDescriptions.html#ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD

File: app/src/main/java/com/framgia/music/SongDataHelper.java [hunk 17]
```diff
@@ -0,0 +1,132 @@
...
+                    song.put("file_name", file.getName());
+                    fileList.add(song);
+                }
+            }
+            
+        } catch (Exception e) {
+            Log.e("Error", "Get list not success ");
+        }
+        finally {
+            return fileList;
```
Comment (le-quang-dao) [inline app/src/main/java/com/framgia/music/SongDataHelper.java]: WARNING: [ReturnFromFinallyBlock] Avoid returning from a finally block
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/basic.html#ReturnFromFinallyBlock

File: app/src/main/res/values/strings.xml [hunk 18]
```diff
@@ -7,4 +7,8 @@
     <string name="add">Add</string>
     <string name="cancel">Cancel</string>
     <string name="input_album_new">Input album new</string>
+    <string name="please_input_album_name">Please input album name</string>
+    <string name="permission_granted">Permission Write File is Granted</string>
```
Comment (le-quang-dao) [inline app/src/main/res/values/strings.xml]: WARNING: [UnusedResources] The resource `R.string.permission_granted` appears to be unused
 Unused resources make applications larger and slow down builds.

File: app/src/main/res/values/strings.xml [hunk 19]
```diff
@@ -7,4 +7,8 @@
...
+    <string name="permission_denied">Permission Write File is Denied</string>
```
Comment (le-quang-dao) [inline app/src/main/res/values/strings.xml]: WARNING: [UnusedResources] The resource `R.string.permission_denied` appears to be unused
 Unused resources make applications larger and slow down builds.

File: app/src/main/res/values/strings.xml [hunk 20]
```diff
@@ -7,4 +7,8 @@
...
+    <string name="permission_show_again">Permisson dont granted and dont show dialog again</string>
```
Comment (le-quang-dao) [inline app/src/main/res/values/strings.xml]: WARNING: [UnusedResources] The resource `R.string.permission_show_again` appears to be unused
 Unused resources make applications larger and slow down builds.

File: app/src/main/java/com/framgia/music/SongDataHelper.java [hunk 21]
```diff
@@ -0,0 +1,129 @@
...
+                    HashMap<String, String> song = new HashMap<>();
+                    song.put("file_path", file.getAbsolutePath());
+                    song.put("file_name", file.getName());
+                    fileList.add(song);
+                }
+            }
+        } catch (Exception e) {
+            throw e;
+        } finally {
+            return fileList;
```
Comment (le-quang-dao) [inline app/src/main/java/com/framgia/music/SongDataHelper.java]: WARNING: [ReturnFromFinallyBlock] Avoid returning from a finally block
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/basic.html#ReturnFromFinallyBlock

File: app/src/main/java/com/framgia/music/AdapterSong.java [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+/**
+ * Created by trant on 29/01/2018.
+ */
+public class AdapterSong extends RecyclerView.Adapter<AdapterSong.ViewHolder> {
+    private ArrayList<SongMusic> mSongs;
+    private Context context;
+
+    public AdapterSong(ArrayList<SongMusic> songMusics, Context context) {
+        this.mSongs = songMusics;
+        this.context = context;
```
Comment (Tuanlvt) [inline app/src/main/java/com/framgia/music/AdapterSong.java]: đưa context lên đầu tiên

File: app/src/main/java/com/framgia/music/AdapterSong.java [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+import android.widget.ImageView;
+import android.widget.TextView;
+import java.util.ArrayList;
+
+/**
+ * Created by trant on 29/01/2018.
+ */
+public class AdapterSong extends RecyclerView.Adapter<AdapterSong.ViewHolder> {
+    private ArrayList<SongMusic> mSongs;
+    private Context context;
```
Comment (Tuanlvt) [inline app/src/main/java/com/framgia/music/AdapterSong.java]: context -> mContext

File: app/src/main/java/com/framgia/music/FragmentSong.java [hunk 24]
```diff
@@ -2,19 +2,47 @@
...
+    private ArrayList<SongMusic> mArraySong;
+    private AdapterSong mAdapterSong;
+    private SongDataHelper mSongDataHelper;
     @Override
     public View onCreateView(LayoutInflater inflater, ViewGroup container,
             Bundle savedInstanceState) {
-        return inflater.inflate(R.layout.fragment_song, container, false);
+        View view = inflater.inflate(R.layout.fragment_song, container, false);
+        mRecyclerSong = (RecyclerView) view.findViewById(R.id.recyclerSong);
+        mSongDataHelper = new SongDataHelper(getContext(), "Data.db", null, 1);
```
Comment (Tuanlvt) [inline app/src/main/java/com/framgia/music/FragmentSong.java]: "Data.db"  -- hardcode

File: app/src/main/java/com/framgia/music/MainActivity.java [hunk 25]
```diff
@@ -42,5 +49,27 @@ public boolean onOptionsItemSelected(MenuItem item) {
...
+    }
+
+    private void initPermission() {
+        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE)
+                != PackageManager.PERMISSION_GRANTED) {
+            ActivityCompat.requestPermissions(this,
+                    new String[] { Manifest.permission.WRITE_EXTERNAL_STORAGE },
+                    REQUEST_PERMISSION);
+        } else {
+            mSongDataHelper = new SongDataHelper(this, "Data.db", null, 1);
```
Comment (Tuanlvt) [inline app/src/main/java/com/framgia/music/MainActivity.java]: hardcode (Data.db)

File: app/src/main/java/com/framgia/music/SongMusic.java [hunk 26]
```diff
@@ -5,47 +5,57 @@
...
-        return favorite;
+        return mFavorite;
     }
 
     public void setFavorite(int favorite) {
-        this.favorite = favorite;
+        mFavorite = favorite;
     }
-}
+}
```
Comment (Tuanlvt) [inline app/src/main/java/com/framgia/music/SongMusic.java]: enter end line

File: app/src/main/res/layout/activity_main.xml [hunk 27]
```diff
@@ -109,7 +109,7 @@
             android:layout_height="@dimen/dp_50"
             android:contentDescription="@string/description"
             android:layout_weight="1"
-            app:srcCompat="@drawable/favorite"
+            app:srcCompat="@drawable/unfavorite"
```
Comment (Tuanlvt) [inline app/src/main/res/layout/activity_main.xml]: unfavorite -> ic_unFavorite

File: app/src/main/java/com/framgia/music/AdapterSong.java (hunk 23 above)
Comment (ThuanPx) [inline app/src/main/java/com/framgia/music/AdapterSong.java]: sai convention

File: app/src/main/java/com/framgia/music/AdapterSong.java [hunk 28]
```diff
@@ -0,0 +1,67 @@
...
+        holder.setData(position);
+    }
+
+    @Override
+    public int getItemCount() {
+        return mSongs.size();
+    }
+
+    public class ViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
+        TextView mTxtNameSong;
```
Comment (ThuanPx) [inline app/src/main/java/com/framgia/music/AdapterSong.java]: private vào

File: app/src/main/java/com/framgia/music/AdapterSong.java [hunk 29]
```diff
@@ -0,0 +1,67 @@
...
+            mTxtNameSong = (TextView) itemView.findViewById(R.id.txtNameRowSong);
+            mImgFavorite = (ImageView) itemView.findViewById(R.id.btnFavRowSong);
+            mUnImgFavorite = (ImageView) itemView.findViewById(R.id.btnUnFavRowSong);
+            mImgDelete = (ImageView) itemView.findViewById(R.id.btndelRowSong);
+            mImgAdd = (ImageView) itemView.findViewById(R.id.btnAddRowSong);
+        }
+
+        public void setData(final int pos) {
+            mTxtNameSong.setText(mSongs.get(pos).getName());
+            mImgDelete.setOnClickListener(this);
```
Comment (ThuanPx) [inline app/src/main/java/com/framgia/music/AdapterSong.java]: chuyển qua dùng callback đi e

File: app/src/main/java/com/framgia/music/AlbumDataHelper.java [hunk 30]
```diff
@@ -45,9 +45,8 @@ public void insertAlbumIntoTableAlbum(String nameAlbum) {
         Cursor c = sqLiteDatabase.query("TableAlbum", null, null, null, null, null, null);
         c.moveToFirst();
         while (!c.isAfterLast()) {
-            int id = c.getInt(0);
-            String name = c.getString(1);
-            //c.getString(2);
+            int id = c.getInt(c.getColumnIndex("idAlbum"));
```
Comment (ThuanPx) [inline app/src/main/java/com/framgia/music/AlbumDataHelper.java]: hard code

File: app/src/main/java/com/framgia/music/SongDataHelper.java [hunk 31]
```diff
@@ -0,0 +1,129 @@
...
+        SQLiteDatabase sqLiteDatabase = this.getReadableDatabase();
+        ArrayList<SongMusic> songMusics = new ArrayList<>();
+        Cursor c = sqLiteDatabase.query("TableSong", null, null, null, null, null, null);
+        c.moveToFirst();
+        while (!c.isAfterLast()) {
+            int id = c.getInt(c.getColumnIndex("idSong"));
+            String name = c.getString(c.getColumnIndex("nameSong"));
+            String path = c.getString(c.getColumnIndex("pathSong"));
+            int favorite = c.getInt(c.getColumnIndex("favorite"));
+            int idAlbum = c.getInt(c.getColumnIndex("idAlbum"));
```
Comment (Tuanlvt) [inline app/src/main/java/com/framgia/music/SongDataHelper.java]: có thể chuyển nhưng name column thành 1 hằng ký tự để gọi ra

File: app/src/main/java/com/framgia/music/SongDataHelper.java [hunk 32]
```diff
@@ -0,0 +1,129 @@
...
+public class SongDataHelper extends SQLiteOpenHelper {
+
+    public SongDataHelper(Context context, String name, SQLiteDatabase.CursorFactory factory,
+            int version) {
+        super(context, name, factory, version);
+    }
+
+    @Override
+    public void onCreate(SQLiteDatabase sqLiteDatabase) {
+        String sql = "CREATE TABLE TableSong ("
```
Comment (ThuanPx) [inline app/src/main/java/com/framgia/music/SongDataHelper.java]: hard code quá, mấy tên table e nên khai báo thành biến

File: app/src/main/java/com/framgia/music/SongDataHelper.java [hunk 33]
```diff
@@ -0,0 +1,129 @@
...
+                    insertSongToDatabase(songListStorage.get(i).get("file_name"), pathSong);
+                }
+            }
+        }
+    }
+
+    public void insertSongToDatabase(String fileName, String pathSong) {
+        SQLiteDatabase db = getWritableDatabase();
+        ContentValues values = new ContentValues();
+        values.put("nameSong", fileName);
```
Comment (ThuanPx) [inline app/src/main/java/com/framgia/music/SongDataHelper.java]: same

File: app/src/main/java/com/framgia/music/SongDataHelper.java [hunk 34]
```diff
@@ -0,0 +1,129 @@
...
+        values.put("pathSong", pathSong);
+        values.put("idAlbum", "");
+        values.put("favorite", 0);
```
Comment (Tuanlvt) [inline app/src/main/java/com/framgia/music/SongDataHelper.java]: có thể chuyển nhưng name column thành 1 hằng ký tự để gọi ra

File: app/src/main/java/com/framgia/music/AdapterAlbum.java [hunk 35]
```diff
@@ -31,7 +37,7 @@ public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
 
     @Override
     public void onBindViewHolder(ViewHolder holder, int position) {
-        holder.setData(position);
+        holder.setData(position, mListenner);
```
Comment (ThuanPx) [inline app/src/main/java/com/framgia/music/AdapterAlbum.java]: mListener nên set ở new ViewHolder(itemView) ấy

File: app/src/main/java/com/framgia/music/AdapterSong.java [hunk 36]
```diff
@@ -0,0 +1,90 @@
...
+    @Override
+    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
+        LayoutInflater layoutInflater = LayoutInflater.from(parent.getContext());
+        View itemView = layoutInflater.inflate(R.layout.row_song, parent, false);
+        return new ViewHolder(itemView);
+    }
+
+    @Override
+    public void onBindViewHolder(ViewHolder holder, int position) {
+        holder.setData(position, mListenner);
```
Comment (ThuanPx) [inline app/src/main/java/com/framgia/music/AdapterSong.java]: same

File: app/src/main/java/com/framgia/music/FragmentSong.java [hunk 37]
```diff
@@ -2,19 +2,61 @@
...
+                //TODO
+                break;
+            case R.id.btnAddRowSong:
+                //TODO
+                break;
+        }
     }
 }
 
+
```
Comment (ThuanPx) [inline app/src/main/java/com/framgia/music/FragmentSong.java]: remove black line

File: app/src/main/java/com/framgia/music/OnItemClickListenner.java [hunk 38]
```diff
@@ -0,0 +1,11 @@
...
+
+import android.view.View;
+
+/**
+ * Created by trant on 31/01/2018.
+ */
+
+public interface OnItemClickListenner {
+    void onClick(int position, View view);
+}
```
Comment (ThuanPx) [inline app/src/main/java/com/framgia/music/OnItemClickListenner.java]: enter

File: app/src/main/java/com/framgia/music/AdapterAlbum.java [hunk 39]
```diff
@@ -43,27 +49,30 @@ public int getItemCount() {
...
+        public ViewHolder(View itemView, OnItemClickListenner listenner) {
             super(itemView);
+            mListenner = listenner;
             mTxtNameAlbum = (TextView) itemView.findViewById(R.id.txtNameRowAlbum);
             mImgEdit = (ImageView) itemView.findViewById(R.id.btnEditRowAlbum);
             mImgDelete = (ImageView) itemView.findViewById(R.id.btnDeleteRowAlbum);
         }
 
-        public void setData(int pos) {
+        public void setData(final int pos) {
```
Comment (ThuanPx) [inline app/src/main/java/com/framgia/music/AdapterAlbum.java]: không cần final cũng đc mà e nhỉ

File: app/src/main/java/com/framgia/music/AdapterAlbum.java [hunk 40]
```diff
@@ -43,27 +49,30 @@ public int getItemCount() {
...
             mTxtNameAlbum.setText(mAlbums.get(pos).getNameAlbum());
-            mImgEdit.setOnClickListener(new View.OnClickListener() {
+            mImgDelete.setOnClickListener(new View.OnClickListener() {
                 @Override
                 public void onClick(View view) {
-                    //TODO
+                    mListenner.onClick(getAdapterPosition(), view);
```
Comment (ThuanPx) [inline app/src/main/java/com/framgia/music/AdapterAlbum.java]: getAdapterPosition() = pos cũng đc nè e

File: app/src/main/java/com/framgia/music/AdapterAlbum.java [hunk 41]
```diff
@@ -43,27 +49,30 @@ public int getItemCount() {
...
         }
 
         public void setData(int pos) {
             mTxtNameAlbum.setText(mAlbums.get(pos).getNameAlbum());
-            mImgEdit.setOnClickListener(new View.OnClickListener() {
+            mImgDelete.setOnClickListener(new View.OnClickListener() {
                 @Override
                 public void onClick(View view) {
-                    //TODO
+                    mListenner.onClick( view);
```
Comment (le-quang-dao) [inline app/src/main/java/com/framgia/music/AdapterAlbum.java]: ERROR: '(' is followed by whitespace.

File: app/src/main/java/com/framgia/music/AdapterAlbum.java [hunk 42]
```diff
@@ -43,27 +49,30 @@ public int getItemCount() {
...
                 }
             });
-            mImgDelete.setOnClickListener(new View.OnClickListener() {
+            mImgEdit.setOnClickListener(new View.OnClickListener() {
                 @Override
                 public void onClick(View view) {
-                    //TODO
+                    mListenner.onClick( view);
```
Comment (le-quang-dao) [inline app/src/main/java/com/framgia/music/AdapterAlbum.java]: ERROR: '(' is followed by whitespace.

File: app/src/main/java/com/framgia/music/AdapterSong.java [hunk 43]
```diff
@@ -0,0 +1,92 @@
...
+            mImgAdd = (ImageView) itemView.findViewById(R.id.btnAddRowSong);
+        }
+
+        public void setData(int pos) {
+
+            mTxtNameSong.setText(mSongs.get(pos).getName());
+            mImgDelete.setOnClickListener(new View.OnClickListener() {
+                @Override
+                public void onClick(View view) {
+                    mListenner.onClick( view);
```
Comment (le-quang-dao) [inline app/src/main/java/com/framgia/music/AdapterSong.java]: ERROR: '(' is followed by whitespace.

**Pull request merged**